client/src/main/java/io/pravega/client/stream/impl/ControllerResolverFactory.java

```diff
@@ -172,32 +171,38 @@ public void start(Listener listener) {
             Preconditions.checkState(this.resolverUpdater == null, "ControllerNameResolver has already been started");
             Preconditions.checkState(!shutdown, "ControllerNameResolver is shutdown, restart is not supported");
             this.resolverUpdater = listener;
-
+            boolean scheduleDiscovery;
             // If the servers comprise only of IP addresses then we need to update the controller list only once.
-            if (this.scheduledExecutor == null) {
+            List<EquivalentAddressGroup> servers = new ArrayList<>();
+            if (!this.enableDiscovery) {
+                scheduleDiscovery = false;
                 // Use the bootstrapped server list as the final set of controllers.
-                List<EquivalentAddressGroup> servers = this.bootstrapServers.stream()
-                        .map(address -> new EquivalentAddressGroup(
-                                new InetSocketAddress(address.getHostString(), address.getPort())))
-                        .collect(Collectors.toList());
+                for (InetSocketAddress address : bootstrapServers) {
+                    if (InetAddresses.isInetAddress(address.getHostString())) {
+                        servers.add(new EquivalentAddressGroup(
+                                new InetSocketAddress(address.getHostString(), address.getPort())));
+                    } else {
+                        scheduleDiscovery = true;
+                    }
+                }
+            } else {
+                scheduleDiscovery = true;
+            }
+            if (scheduleDiscovery) {
+                // Schedule the first discovery immediately.
+                this.scheduledFuture = this.scheduledExecutor.schedule(this::getControllers, 0L, TimeUnit.SECONDS);
```

Comment:
- If the `bootstrap servers` do not contain an IP address then it is scheduled for discovery/resolution. The `getControllers()` method uses `java.net.InetSocketAddress` to resolve it. Now the default value of system property `networkaddress.cache.ttl` is -1. i.e. all addresses once resolved will be cached forever. This would defeat the goal of getControllers() being invoked periodically.
https://docs.oracle.com/javase/8/docs/technotes/guides/net/properties.html#nct

- If the dns resolution takes more time when we are using the `io.pravega.client.netty.impl.ConnectionFactory#getInternalExecutor()` to run the resolver task it would affect the other segment readers and writers right ?

Reply:
1) Yes it would... I am not sure what to do about that though. Because this is a library we can't go changing system properties which have global effects.
2) Yes.

Reply:
Point 2 will be problematic to debug.
As a next step let me trigger a round of system tests to verify it.